File: packages/shapes/src/lib/canonical-json.ts
import { Pile } from './domain-types';
import { GameState } from './game-state-types';

/**
 * Canonical serialisation of a `GameState` for cross-implementation comparison.
 *
 * Two implementations agree iff their canonical strings are byte-identical, so every
 * rule here is part of the contract with the Kotlin port and is documented in
 * `docs/game-engine/RECORDING.md`:
 *
 * - object keys sorted lexicographically at every level
 * - arrays in order; a `Pile` becomes a plain array, top card first
 * - `undefined` properties omitted, `null` kept
 * - no whitespace
 * - integers only — a fractional number throws, because TypeScript prints `1` where
 *   Kotlin prints `1.0` and the two would silently diverge
 */

/** Fields excluded from the canonical form. Everything else in `GameState` is hashed. */
export const CANONICAL_EXCLUDED_STATE_FIELDS = [
  // Client-authored history whose `description` strings are user-facing prose. Hashing
  // them would make UI copy part of the cross-language contract.
  'turnActions',
  'roundActions',
] as const;

/** Per-player fields excluded from the canonical form. */
export const CANONICAL_EXCLUDED_PLAYER_FIELDS = [
  // Bot-internal, contains floats, and never written into GameState by the engine.
  'botMemory',
] as const;

function assertCanonicalNumber(value: number, path: string): void {
  if (!Number.isFinite(value)) {
    throw new TypeError(`Non-finite number at ${path}: ${value}`);
  }
  if (!Number.isInteger(value)) {
    throw new TypeError(
      `Non-integer number at ${path}: ${value}. GameState must contain integers only — ` +
        `TypeScript prints 1 where Kotlin prints 1.0, which would break parity.`,
    );
  }
}

function canonicalize(value: unknown, path: string): string | undefined {
  if (value === undefined) return undefined;
  if (value === null) return 'null';

  if (typeof value === 'number') {
    assertCanonicalNumber(value, path);
    return String(value);
  }

  if (typeof value === 'string') return JSON.stringify(value);
  if (typeof value === 'boolean') return value ? 'true' : 'false';

  if (value instanceof Pile) {
    return canonicalize(value.toArray(), path);
  }

  if (Array.isArray(value)) {
    const items = value.map(
      (item, index) => canonicalize(item, `${path}[${index}]`) ?? 'null',
    );
    return `[${items.join(',')}]`;
  }

  if (typeof value === 'object') {
    const source = value as Record<string, unknown>;
    const entries: string[] = [];

    for (const key of Object.keys(source).sort()) {
      const serialised = canonicalize(source[key], `${path}.${key}`);
      if (serialised !== undefined) {
        entries.push(`${JSON.stringify(key)}:${serialised}`);
      }
    }

    return `{${entries.join(',')}}`;
  }

  throw new TypeError(`Unsupported value at ${path}: ${typeof value}`);
}

/** Strips excluded fields, leaving a plain structure the generic canonicaliser can walk. */
function toCanonicalShape(state: GameState): Record<string, unknown> {
  const shaped: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(state)) {
    if ((CANONICAL_EXCLUDED_STATE_FIELDS as readonly string[]).includes(key)) {
      continue;
    }

    if (key === 'players' && Array.isArray(value)) {
      shaped[key] = value.map((player) => {
        const kept: Record<string, unknown> = {};
        for (const [playerKey, playerValue] of Object.entries(
          player as Record<string, unknown>,
        )) {
          if (
            !(CANONICAL_EXCLUDED_PLAYER_FIELDS as readonly string[]).includes(
              playerKey,
            )
          ) {
            kept[playerKey] = playerValue;
          }
        }
        return kept;
      });
      continue;
    }

    shaped[key] = value;
  }

  return shaped;
}

export function canonicalizeGameState(state: GameState): string {
  return canonicalize(toCanonicalShape(state), '$') ?? 'null';
}

/**
 * Lowercase hex SHA-256 of the canonical string.
 *
 * Async and built on WebCrypto rather than `node:crypto` so there is a single code path
 * in Node, the browser and tests — the same reason the Kotlin side uses one pure-Kotlin
 * SHA-256 across all targets instead of per-platform implementations.
 */
export async function hashCanonicalString(canonical: string): Promise<string> {
  const bytes = new TextEncoder().encode(canonical);
  const digest = await globalThis.crypto.subtle.digest('SHA-256', bytes);

  return Array.from(new Uint8Array(digest))
    .map((byte) => byte.toString(16).padStart(2, '0'))
    .join('');
}

export function hashGameState(state: GameState): Promise<string> {
  return hashCanonicalString(canonicalizeGameState(state));
}

File: kmp/shared/shapes/src/commonMain/kotlin/game/vinto/shapes/CanonicalJson.kt
package game.vinto.shapes

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/**
 * Canonical serialisation of a `GameState` for cross-implementation comparison.
 *
 * Two implementations agree iff their canonical strings are byte-identical, so every rule
 * here is part of the contract with TypeScript and is documented in
 * `docs/game-engine/RECORDING.md`:
 *
 *  - object keys sorted lexicographically at every level
 *  - arrays in order; a `Pile` is a plain array, top card first
 *  - absent properties stay absent, `null` is kept
 *  - no whitespace
 *  - integers only - a fractional number throws, because TypeScript prints `1` where
 *    Kotlin would print `1.0` and the two would diverge silently
 *
 * It walks a [JsonElement] rather than the typed model on purpose. TypeScript distinguishes
 * an absent property from a null one and the canonical form preserves that; JSON already
 * carries that distinction, whereas a Kotlin data class has only `null` for both. Encoding
 * the typed model to JSON first (which [Card] documents how to get right) and canonicalising
 * that keeps the rule in exactly one place.
 */
object CanonicalJson {

    /** Client-authored history whose `description` strings are user-facing prose. */
    val EXCLUDED_STATE_FIELDS = setOf("turnActions", "roundActions")

    /** Bot-internal, contains floats, never written into `GameState` by the engine. */
    val EXCLUDED_PLAYER_FIELDS = setOf("botMemory")

    /** Canonical string of an already-encoded state, applying the documented exclusions. */
    fun ofGameState(state: JsonObject): String = canonicalize(toCanonicalShape(state), "$")

    /** Canonical string of any element, with no field exclusions. */
    fun of(element: JsonElement): String = canonicalize(element, "$")

    private fun toCanonicalShape(state: JsonObject): JsonObject {
        val shaped = LinkedHashMap<String, JsonElement>()

        for ((key, value) in state) {
            if (key in EXCLUDED_STATE_FIELDS) continue

            if (key == "players" && value is JsonArray) {
                shaped[key] = JsonArray(
                    value.map { player ->
                        if (player !is JsonObject) player
                        else JsonObject(player.filterKeys { it !in EXCLUDED_PLAYER_FIELDS })
                    },
                )
                continue
            }

            shaped[key] = value
        }

        return JsonObject(shaped)
    }

    private fun canonicalize(value: JsonElement, path: String): String = when (value) {
        is JsonNull -> "null"
        is JsonPrimitive -> canonicalizePrimitive(value, path)
        is JsonArray -> value
            .mapIndexed { index, item -> canonicalize(item, "$path[$index]") }
            .joinToString(",", "[", "]")

        is JsonObject -> value.keys
            .sorted()
            .joinToString(",", "{", "}") { key ->
                quote(key) + ":" + canonicalize(value.getValue(key), "$path.$key")
            }
    }

    private fun canonicalizePrimitive(value: JsonPrimitive, path: String): String {
        if (value.isString) return quote(value.content)

        return when (val content = value.content) {
            "true", "false" -> content
            else -> {
                val number = content.toDoubleOrNull()
                    ?: throw IllegalArgumentException("Unsupported value at $path: $content")
                if (number.isNaN() || number.isInfinite()) {
                    throw IllegalArgumentException("Non-finite number at $path: $content")
                }
                content.toLongOrNull()?.toString()
                    ?: throw IllegalArgumentException(
                        "Non-integer number at $path: $content. GameState must contain " +
                            "integers only - TypeScript prints 1 where Kotlin prints 1.0, " +
                            "which would break parity.",
                    )
            }
        }
    }

    /**
     * JSON string literal, matching `JSON.stringify`: the two mandatory escapes, the five
     * short forms, `\uXXXX` for the remaining control characters, and everything else -
     * including non-ASCII - emitted raw, which is what JavaScript does.
     */
    private fun quote(text: String): String = buildString(text.length + 2) {
        append('"')
        for (char in text) {
            when (char) {
                '"' -> append("\\\"")
                '\\' -> append("\\\\")
                '\b' -> append("\\b")
                '\u000C' -> append("\\f")
                '\n' -> append("\\n")
                '\r' -> append("\\r")
                '\t' -> append("\\t")
                else -> if (char < ' ') {
                    append("\\u")
                    append(char.code.toString(16).padStart(4, '0'))
                } else {
                    append(char)
                }
            }
        }
        append('"')
    }
}
